agent/consul/leader_connect_ca.go

```diff
@@ -503,39 +510,13 @@ func (c *CAManager) primaryInitialize(provider ca.Provider, conf *structs.CAConf
 		return fmt.Errorf("error generating CA root certificate: %v", err)
 	}
 
-	rootCA, err := newCARoot(root.PEM, conf.Provider, conf.ClusterID)
+	rootCA, err := newCARoot(root, conf.Provider, conf.ClusterID)
 	if err != nil {
 		return err
 	}
 
-	// TODO: https://github.com/hashicorp/consul/issues/12386
-	interPEM, err := provider.GenerateLeafSigningCert()
-	if err != nil {
-		return fmt.Errorf("error generating intermediate cert: %v", err)
-	}
-	intermediateCert, err := connect.ParseCert(interPEM)
-	if err != nil {
-		return fmt.Errorf("error getting intermediate cert: %v", err)
-	}
-
 	var rootUpdateRequired bool
-
-	// Versions prior to 1.9.3, 1.8.8, and 1.7.12 incorrectly used the primary
-	// rootCA's subjectKeyID here instead of the intermediate. For
-	// provider=consul this didn't matter since there are no intermediates in
-	// the primaryDC, but for vault it does matter.
-	expectedSigningKeyID := connect.EncodeSigningKeyID(intermediateCert.SubjectKeyId)
-	if rootCA.SigningKeyID != expectedSigningKeyID {
-		c.logger.Info("Correcting stored CARoot values",
-			"previous-signing-key", rootCA.SigningKeyID, "updated-signing-key", expectedSigningKeyID)
-		rootCA.SigningKeyID = expectedSigningKeyID
-		rootUpdateRequired = true
-	}
-
-	// Add the local leaf signing cert to the rootCA struct. This handles both
-	// upgrades of existing state, and new rootCA.
-	if c.getLeafSigningCertFromRoot(rootCA) != interPEM {
-		rootCA.IntermediateCerts = append(rootCA.IntermediateCerts, interPEM)
```

Comment:
All of this removed logic comes for free in `setLeafSigningCert`.

Reply:
We lose the logging but I'm not sure how much value it has. Anecdotally I was more confused by it when I first encountered it.